// @polkadot/util-crypto/signature/verify.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = signatureVerify;

var _util = require("@polkadot/util");

var _decode = _interopRequireDefault(require("../address/decode"));

var _verify = _interopRequireDefault(require("../nacl/verify"));

var _verify2 = _interopRequireDefault(require("../schnorrkel/verify"));

var _verify3 = _interopRequireDefault(require("../secp256k1/verify"));

// Copyright 2017-2020 @polkadot/util-crypto authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
const VERIFIERS = [['ed25519', _verify.default], ['sr25519', _verify2.default], ['ecdsa', _verify3.default]];
const CRYPTO_TYPES = ['ed25519', 'sr25519', 'ecdsa'];

function verifyDetect(result, message, signature, publicKey) {
  result.isValid = VERIFIERS.some(([crypto, verify]) => {
    try {
      if (verify(message, signature, publicKey)) {
        result.crypto = crypto;
        return true;
      }
    } catch (error) {// do nothing, result.isValid still set to false
    }

    return false;
  });
  return result;
}

function verifyMultisig(result, message, signature, publicKey) {
  (0, _util.assert)([0, 1, 2].includes(signature[0]), `Unknown crypto type, expected signature prefix [0..2], found ${signature[0]}`);
  result.crypto = CRYPTO_TYPES[signature[0]] || 'none';

  try {
    result.isValid = {
      ecdsa: () => (0, _verify3.default)(message, signature.subarray(1), publicKey),
      ed25519: () => (0, _verify.default)(message, signature.subarray(1), publicKey),
      none: () => {
        throw Error('no verify for `none` crypto type');
      },
      sr25519: () => (0, _verify2.default)(message, signature.subarray(1), publicKey)
    }[result.crypto]();
  } catch (error) {// ignore, result.isValid still set to false
  }

  return result;
}

function signatureVerify(message, signature, addressOrPublicKey) {
  const signatureU8a = (0, _util.u8aToU8a)(signature);
  (0, _util.assert)([64, 65, 66].includes(signatureU8a.length), `Invalid signature length, expected [64..66] bytes, found ${signatureU8a.length}`);
  const result = {
    crypto: 'none',
    isValid: false
  };
  const publicKey = (0, _decode.default)(addressOrPublicKey);
  const multisig = [0, 1, 2].includes(signatureU8a[0]) && [65, 66].includes(signatureU8a.length);

  if (multisig) {
    return verifyMultisig(result, message, signatureU8a, publicKey);
  } else {
    return verifyDetect(result, message, signatureU8a, publicKey);
  }
}
