// @polkadot/types/generic/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GenericAccountId", {
  enumerable: true,
  get: function () {
    return _AccountId.default;
  }
});
Object.defineProperty(exports, "GenericAccountIndex", {
  enumerable: true,
  get: function () {
    return _AccountIndex.default;
  }
});
Object.defineProperty(exports, "GenericAddress", {
  enumerable: true,
  get: function () {
    return _Address.default;
  }
});
Object.defineProperty(exports, "GenericBlock", {
  enumerable: true,
  get: function () {
    return _Block.default;
  }
});
Object.defineProperty(exports, "GenericCall", {
  enumerable: true,
  get: function () {
    return _Call.default;
  }
});
Object.defineProperty(exports, "GenericConsensusEngineId", {
  enumerable: true,
  get: function () {
    return _ConsensusEngineId.default;
  }
});
Object.defineProperty(exports, "GenericEvent", {
  enumerable: true,
  get: function () {
    return _Event.default;
  }
});
Object.defineProperty(exports, "GenericVote", {
  enumerable: true,
  get: function () {
    return _Vote.default;
  }
});

var _AccountId = _interopRequireDefault(require("./AccountId"));

var _AccountIndex = _interopRequireDefault(require("./AccountIndex"));

var _Address = _interopRequireDefault(require("./Address"));

var _Block = _interopRequireDefault(require("./Block"));

var _Call = _interopRequireDefault(require("./Call"));

var _ConsensusEngineId = _interopRequireDefault(require("./ConsensusEngineId"));

var _Event = _interopRequireDefault(require("./Event"));

var _Vote = _interopRequireDefault(require("./Vote"));
