// prisma-db-introspection/dist/common/introspectionResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prisma_datamodel_1 = require("prisma-datamodel");
var defaultNormalizer_1 = require("./normalization/defaultNormalizer");
var IntrospectionResult = /** @class */ (function () {
    function IntrospectionResult(databaseType, renderer) {
        this.renderer = renderer || prisma_datamodel_1.DefaultRenderer.create(databaseType);
        this.databaseType = databaseType;
    }
    IntrospectionResult.prototype.renderToDatamodelString = function () {
        return this.renderer.render(this.getDatamodel(), true);
    };
    IntrospectionResult.prototype.getNormalizedDatamodel = function (baseModel) {
        if (baseModel === void 0) { baseModel = null; }
        var model = this.getDatamodel();
        defaultNormalizer_1.default.create(this.databaseType, baseModel).normalize(model);
        return model;
    };
    /**
     * Performs name normalization and order normalization.
     *
     * If a base model is given, order and additional directives are taken
     * from the base model.
     * @param baseModel
     */
    IntrospectionResult.prototype.renderToNormalizedDatamodelString = function (baseModel) {
        if (baseModel === void 0) { baseModel = null; }
        return this.renderer.render(this.getNormalizedDatamodel(baseModel));
    };
    return IntrospectionResult;
}());
exports.IntrospectionResult = IntrospectionResult;
