// prisma-generate-schema/dist/src/generator/default/mutation/create/modelCreateInputGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../../../generator");
var prisma_datamodel_1 = require("prisma-datamodel");
var ModelCreateInputGenerator = /** @class */ (function (_super) {
    __extends(ModelCreateInputGenerator, _super);
    function ModelCreateInputGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Generates an create model input field for a relational type, handling the four cases many/one and with/without related type.
     * @param model
     * @param field
     * @param generators
     */
    ModelCreateInputGenerator.generateRelationFieldForInputType = function (model, field, generators) {
        var relationInfo = {
            relatedField: field,
            relatedType: model,
            relationName: field.relationName,
        };
        var generator = ModelCreateInputGenerator.getGeneratorForRelation(model, field, generators);
        if (generator.wouldBeEmpty(field.type, relationInfo)) {
            return null;
        }
        else {
            var schemaField = generator.generate(field.type, relationInfo);
            return generators.scalarTypeGenerator.requiredIf(field.isRequired, schemaField);
        }
    };
    ModelCreateInputGenerator.relationWouldBeEmpty = function (model, field, generators) {
        var relationInfo = {
            relatedField: field,
            relatedType: model,
            relationName: field.relationName,
        };
        var generator = ModelCreateInputGenerator.getGeneratorForRelation(model, field, generators);
        return generator.wouldBeEmpty(field.type, relationInfo);
    };
    ModelCreateInputGenerator.getGeneratorForRelation = function (model, field, generators) {
        if (field.relatedField !== null) {
            if (field.isList) {
                return generators.modelCreateManyWithoutRelatedInput;
            }
            else {
                return generators.modelCreateOneWithoutRelatedInput;
            }
        }
        else {
            if (field.isList) {
                return generators.modelCreateManyInput;
            }
            else {
                return generators.modelCreateOneInput;
            }
        }
    };
    // TODO: Make all those instance members.
    ModelCreateInputGenerator.generateScalarFieldTypeForInputType = function (model, field, generators) {
        if ((field.isReadOnly &&
            !(field.idStrategy === prisma_datamodel_1.IdStrategy.Auto ||
                field.idStrategy === prisma_datamodel_1.IdStrategy.None)) ||
            field.isCreatedAt ||
            field.isUpdatedAt) {
            return null;
        }
        else {
            if (field.isList) {
                return generators.scalarListCreateInput.generate(model, field);
            }
            else {
                return generators.scalarTypeGenerator.mapToScalarFieldTypeForInput(field);
            }
        }
    };
    ModelCreateInputGenerator.prototype.wouldBeEmpty = function (model, args) {
        var _this = this;
        return (!this.hasCreateInputFields(this.getScalarFields(model.fields)) &&
            this.getRelationFields(model.fields).every(function (field) {
                return ModelCreateInputGenerator.relationWouldBeEmpty(model, field, _this.generators);
            }));
    };
    ModelCreateInputGenerator.prototype.getTypeName = function (input, args) {
        return input.name + "CreateInput";
    };
    /**
     * Generates all fields of this type.
     * @param model
     * @param args
     */
    ModelCreateInputGenerator.prototype.generateFields = function (model, args) {
        var fields = {};
        var fieldList = this.getCreateInputFields(model.fields);
        for (var _i = 0, fieldList_1 = fieldList; _i < fieldList_1.length; _i++) {
            var field = fieldList_1[_i];
            var isScalar = this.generators.scalarTypeGenerator.isScalarField(field);
            var fieldSchema = isScalar
                ? this.generateScalarField(model, args, field)
                : this.generateRelationField(model, args, field);
            if (fieldSchema !== null) {
                fields[field.name] = fieldSchema;
            }
        }
        return fields;
    };
    ModelCreateInputGenerator.prototype.generateRelationFieldType = function (model, args, field) {
        return ModelCreateInputGenerator.generateRelationFieldForInputType(model, field, this.generators);
    };
    ModelCreateInputGenerator.prototype.generateScalarFieldType = function (model, args, field) {
        return ModelCreateInputGenerator.generateScalarFieldTypeForInputType(model, field, this.generators);
    };
    return ModelCreateInputGenerator;
}(generator_1.ModelInputObjectTypeGenerator));
exports.default = ModelCreateInputGenerator;
