// @polkadot/types-known/modules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
// Copyright 2017-2020 @polkadot/types-known authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
// type overrides for modules (where duplication between modules exist)
const typesModules = {
  balances: {
    Releases: 'ReleasesBalances'
  },
  contract: {
    // old metadata & naming
    // v2 & v3
    AccountInfo: 'ContractAccountInfo'
  },
  contracts: {
    StorageKey: 'ContractStorageKey'
  },
  democracy: {
    Releases: 'ReleasesDemocracy'
  },
  identity: {
    Judgement: 'IdentityJudgement'
  },
  parachains: {
    Id: 'ParaId'
  },
  society: {
    Judgement: 'SocietyJudgement',
    Vote: 'SocietyVote'
  },
  staking: {
    Compact: 'CompactAssignments',
    Releases: 'ReleasesStaking'
  },
  treasury: {
    Proposal: 'TreasuryProposal'
  }
};
var _default = typesModules;
exports.default = _default;
