// prisma-db-introspection/dist/common/normalization/modelNameAndDirectiveNormalizer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var prisma_datamodel_1 = require("prisma-datamodel");
var modelNameNormalizer_1 = require("./modelNameNormalizer");
var ModelNameAndDirectiveNormalizer = /** @class */ (function (_super) {
    __extends(ModelNameAndDirectiveNormalizer, _super);
    function ModelNameAndDirectiveNormalizer(baseModel) {
        return _super.call(this, baseModel) || this;
    }
    ModelNameAndDirectiveNormalizer.prototype.findAndSetBaseByRelation = function (obj) {
        var _this = this;
        if (this.baseFields === null || typeof obj.type === 'string') {
            return null;
        }
        var fieldType = obj.type;
        var _a = __read(this.baseFields.filter(function (base) {
            return (typeof base.type !== 'string' &&
                base.type.name === fieldType.name &&
                // Only check the relation name if is it set on base. It might have been omitted.
                (base.relationName === null ||
                    obj.relationName === null ||
                    base.relationName === obj.relationName));
        }), 1), baseCandidate = _a[0];
        if (baseCandidate !== undefined) {
            // Use each base field only once, see base class.
            this.baseField = baseCandidate;
            this.baseFields = this.baseFields.filter(function (f) { return _this.baseField !== f; });
            return this.baseField;
        }
        else {
            return null;
        }
    };
    ModelNameAndDirectiveNormalizer.prototype.findAndSetBaseById = function (obj) {
        var _this = this;
        if (this.baseFields === null) {
            return null;
        }
        var _a = __read(this.baseFields.filter(function (baseObj) { return baseObj.isId && obj.isId; }), 1), baseCandidate = _a[0];
        if (baseCandidate !== undefined) {
            // Use each base field only once, see base class.
            this.baseField = baseCandidate;
            this.baseFields = this.baseFields.filter(function (f) { return _this.baseField !== f; });
            return this.baseField;
        }
        else {
            return null;
        }
    };
    ModelNameAndDirectiveNormalizer.prototype.assignTypeProperties = function (baseObj, obj) {
        if (baseObj === null) {
            return;
        }
        if (baseObj.databaseName) {
            obj.name = baseObj.name;
            obj.databaseName = baseObj.databaseName;
        }
    };
    /**
     * Assigns all properties from a base model field
     * to the current field, handling several special cases.
     * @param baseObj The base (reference) field.
     * @param obj The current field.
     * @param parentModel The whole model, for finding enums.
     */
    ModelNameAndDirectiveNormalizer.prototype.assignFieldProperties = function (baseObj, obj, parentModel) {
        if (baseObj === null) {
            return;
        }
        obj.isId = obj.isId || baseObj.isId;
        // In case the field is an ID field, we rename
        // even if there is no database name. This case
        // can happen with mongo, and prisma will be able to remap it.
        if (baseObj.databaseName || obj.isId) {
            obj.name = baseObj.name;
            obj.databaseName = baseObj.databaseName;
        }
        obj.isCreatedAt = obj.isCreatedAt || baseObj.isCreatedAt;
        obj.isUpdatedAt = obj.isUpdatedAt || baseObj.isUpdatedAt;
        obj.defaultValue = obj.defaultValue || baseObj.defaultValue;
        obj.associatedSequence =
            obj.associatedSequence || baseObj.associatedSequence;
        obj.relationName = obj.relationName || baseObj.relationName;
        // Overwrite strategy, if database has none, prisma might have some.
        if (obj.idStrategy === prisma_datamodel_1.IdStrategy.None) {
            obj.idStrategy = baseObj.idStrategy;
        }
        if (baseObj.associatedSequence) {
            obj.associatedSequence = __assign({}, baseObj.associatedSequence);
        }
        // Override JSON mapped to string.
        if (obj.type === prisma_datamodel_1.TypeIdentifiers.string &&
            baseObj.type === prisma_datamodel_1.TypeIdentifiers.json) {
            obj.type = prisma_datamodel_1.TypeIdentifiers.json;
        }
        if (obj.type === prisma_datamodel_1.TypeIdentifiers.string &&
            typeof baseObj.type !== 'string' &&
            baseObj.type.isEnum) {
            // We found an enum type field shadowed by prisma.
            var baseEnumType_1 = baseObj.type;
            // Attempt to find the enum type
            var candidateEnum = parentModel.types.find(function (x) { return x.isEnum && x.name === baseEnumType_1.name; });
            if (candidateEnum !== undefined) {
                obj.type = candidateEnum;
            }
        }
    };
    ModelNameAndDirectiveNormalizer.prototype.normalizeType = function (type, parentModel) {
        this.assignTypeProperties(this.baseType, type);
        _super.prototype.normalizeType.call(this, type, parentModel, this.baseType !== null);
    };
    /**
     * Normalizes a field, handling several special cases.
     * @param baseObj The base (reference) field.
     * @param obj The current field.
     * @param parentModel The whole model, for finding enums.
     */
    ModelNameAndDirectiveNormalizer.prototype.normalizeField = function (field, parentType, parentModel) {
        if (this.baseField !== null) {
            // Base case, we identified a field by name.
            this.assignName(field, this.baseField.name);
            this.assignFieldProperties(this.baseField, field, parentModel);
        }
        else {
            // Fallback to finding a base field by ID.
            this.baseField = this.findAndSetBaseById(field);
            if (this.baseField !== null) {
                this.assignFieldProperties(this.baseField, field, parentModel);
            }
            else {
                // Fallback to finding a base field by relation type
                // Here, we have a few special cases when overwriting properties.
                this.baseField = this.findAndSetBaseByRelation(field);
                if (this.baseField !== null) {
                    // Hard-override name. Relation names are usually auto-generated.
                    field.name = this.baseField.name;
                    field.databaseName = this.baseField.databaseName;
                    if (this.baseField.relationName === null ||
                        field.relationName === null) {
                        // Remove relation name if it is unset in ref model,
                        // Set relation name if set on ref model but not for us.
                        field.relationName = this.baseField.relationName;
                    }
                    // If this is a self-referencing field with a back connection on the same type, we copy
                    // the name of the related field as well. Otherwise, we always
                    // end up overwriting our name with the name of the
                    // first field of same type in the reference model.
                    if (field.type == parentType &&
                        this.baseField.relatedField !== null &&
                        field.relatedField !== null) {
                        field.relatedField.name = this.baseField.relatedField.name;
                        field.relatedField.relationName = this.baseField.relatedField.relationName;
                    }
                    this.assignFieldProperties(this.baseField, field, parentModel);
                }
                else {
                    // If there is absolutely no base field,
                    // Use our name normalization.
                    _super.prototype.normalizeField.call(this, field, parentType, parentModel);
                }
            }
        }
    };
    return ModelNameAndDirectiveNormalizer;
}(modelNameNormalizer_1.default));
exports.default = ModelNameAndDirectiveNormalizer;
