// prisma-generate-schema/dist/src/generator/default/mutation/update/modelUpdateInputGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../../../generator");
var ModelUpdateInputGenerator = /** @class */ (function (_super) {
    __extends(ModelUpdateInputGenerator, _super);
    function ModelUpdateInputGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ModelUpdateInputGenerator.generateScalarFieldTypeForInputType = function (model, field, generators) {
        if (field.isReadOnly || field.isCreatedAt || field.isUpdatedAt) {
            return null;
        }
        if (field.isList) {
            return generators.scalarListUpdateInput.generate(model, field);
        }
        else {
            return generators.scalarTypeGenerator.mapToScalarFieldTypeForceOptional(field);
        }
    };
    /**
     * Generates an update model input field for a relational type, handling the four cases many/one and with/without related type.
     * @param model
     * @param field
     * @param generators
     */
    ModelUpdateInputGenerator.generateRelationFieldForInputType = function (model, field, generators) {
        var relationInfo = {
            relatedField: field,
            relatedType: model,
            relationName: field.relationName,
        };
        var generator = ModelUpdateInputGenerator.getGeneratorForRelation(model, field, generators);
        if (generator.wouldBeEmpty(field.type, relationInfo)) {
            return null;
        }
        else {
            return generator.generate(field.type, relationInfo);
        }
    };
    ModelUpdateInputGenerator.relationWouldBeEmpty = function (model, field, generators) {
        var relationInfo = {
            relatedField: field,
            relatedType: model,
            relationName: field.relationName,
        };
        var generator = ModelUpdateInputGenerator.getGeneratorForRelation(model, field, generators);
        return generator.wouldBeEmpty(field.type, relationInfo);
    };
    ModelUpdateInputGenerator.getGeneratorForRelation = function (model, field, generators) {
        if (field.relatedField !== null) {
            if (field.isList) {
                return generators.modelUpdateManyWithoutRelatedInput;
            }
            else {
                if (field.isRequired) {
                    return generators.modelUpdateOneRequiredWithoutRelatedInput;
                }
                else {
                    return generators.modelUpdateOneWithoutRelatedInput;
                }
            }
        }
        else {
            if (field.isList) {
                return generators.modelUpdateManyInput;
            }
            else {
                if (field.isRequired) {
                    return generators.modelUpdateOneRequiredInput;
                }
                else {
                    return generators.modelUpdateOneInput;
                }
            }
        }
    };
    ModelUpdateInputGenerator.prototype.wouldBeEmpty = function (model, args) {
        return !this.hasWriteableFields(model.fields);
        // TODO: We should add the following check, but this requires
        // a caching mechanism to avoid recursive checks.
        // this.getRelationFields(model.fields).every(field => ModelUpdateInputGenerator.relationWouldBeEmpty(model, field, this.generators))
    };
    ModelUpdateInputGenerator.prototype.getTypeName = function (input, args) {
        return input.name + "UpdateInput";
    };
    ModelUpdateInputGenerator.prototype.generateScalarFieldType = function (model, args, field) {
        return ModelUpdateInputGenerator.generateScalarFieldTypeForInputType(model, field, this.generators);
    };
    ModelUpdateInputGenerator.prototype.generateRelationFieldType = function (model, args, field) {
        return ModelUpdateInputGenerator.generateRelationFieldForInputType(model, field, this.generators);
    };
    return ModelUpdateInputGenerator;
}(generator_1.RelatedModelInputObjectTypeGenerator));
exports.default = ModelUpdateInputGenerator;
